db/revision_cache_lru.go

```diff
@@ -403,12 +438,82 @@ func (value *revCacheValue) store(docRev DocumentRevision) {
 		value.attachments = docRev.Attachments.ShallowCopy() // Don't store attachments the caller might later mutate
 		value.deleted = docRev.Deleted
 		value.err = nil
+		value.itemBytes = docRev.MemoryBytes
 	}
 	value.lock.Unlock()
 }
 
-func (value *revCacheValue) updateDelta(toDelta RevisionDelta) {
+func (value *revCacheValue) updateDelta(toDelta RevisionDelta) (diffInBytes int64) {
 	value.lock.Lock()
+	var previousDeltaBytes int64
+	if value.delta != nil {
+		// delta exists, need to pull this to update overall memory size correctly
+		previousDeltaBytes = value.delta.totalDeltaBytes
+	}
+	diffInBytes = toDelta.totalDeltaBytes - previousDeltaBytes
 	value.delta = &toDelta
+	if diffInBytes != 0 {
+		value.itemBytes += diffInBytes
+	}
 	value.lock.Unlock()
+	return diffInBytes
+}
+
+// getItemBytes retrieves the rev cache items overall memory footprint
+func (value *revCacheValue) getItemBytes() int64 {
+	value.lock.RLock()
+	defer value.lock.RUnlock()
+	return value.itemBytes
+}
+
+// CalculateBytes will calculate the bytes from revisions in the document, body and channels on the document
+func (rev *DocumentRevision) CalculateBytes() {
+	var totalBytes int
+	for v := range rev.Channels {
+		bytes := len([]byte(v))
+		totalBytes += bytes
+	}
+	// calculate history
+	digests, _ := GetStringArrayProperty(rev.History, RevisionsIds)
+	historyBytes := 32 * len(digests)
+	totalBytes += historyBytes
+	// add body bytes into calculation
+	totalBytes += len(rev.BodyBytes)
+
+	// convert the int to int64 type and assign to document revision
+	rev.MemoryBytes = int64(totalBytes)
+}
+
+// CalculateDeltaBytes will calculate bytes from delta channels, delta revisions and delta body
+func (delta *RevisionDelta) CalculateDeltaBytes() {
+	var totalBytes int
+	for v := range delta.ToChannels {
+		bytes := len([]byte(v))
+		totalBytes += bytes
+	}
+	// history calculation
+	historyBytes := 32 * len(delta.RevisionHistory)
+	totalBytes += historyBytes
+
+	// account for delta body
+	totalBytes += len(delta.DeltaBytes)
+
+	delta.totalDeltaBytes = int64(totalBytes)
+}
+
+// revCacheMemoryBasedEviction checks for rev cache eviction, if required calls performEviction which will acquire lock to evict
+func (rc *LRURevisionCache) revCacheMemoryBasedEviction() {
+	// if memory capacity is not set, don't check for eviction this way
+	if rc.memoryCapacity > 0 && rc.memoryBytes.Value() > rc.memoryCapacity {
+		rc.performEviction()
```

Comment:
I'm not sure I follow. This function is to determine if we need eviction based upon if we have a max size set & the current memory count is above the threshold. 
If true it will call `performEviction()` which acquires the lock and runs a loop for eviction till the memory stat is below the threshold.

Reply:
This was my mistake - misread performEviction as 'if' instead of 'for'.